### ISSUE TITLE
Replace JaxrsApplicationScanner by Spring Framework scanning facility

### ISSUE THREAD

**Issue by snicoll** [paypal/resteasy-spring-boot#65]
Is `JaxrsApplicationScanner` scanning the full classpath (not a specific package)? 

If you're using Spring, there is an API to do that, something like:

```java
ClassPathScanningCandidateComponentProvider scanner = new ClassPathScanningCandidateComponentProvider(false);
scanner.addIncludeFilter(new AssignableTypeFilter(Application.class));
Set<BeanDefinition> candidates = scanner.findCandidateComponents(packageToScan);
Set<Class<? extends Application>> classes = new HashSet<Class<? extends Application>>();
for (BeanDefinition candidate : candidates) {
    Class<? extends Application> type = (Class<? extends Application>) ClassUtils.forName(
            candidate.getBeanClassName(), JaxrsApplicationScanner.class.getClassLoader());
    classes.add(type);
}
return classes;
```

It would be more consistent that the actual value of component scan was used for the scanning (by default scan only the package in which the `@SpringBootApplication` is defined

**Comment by fabiocarvalho777** [comment]
Actually, [as documented here](https://github.com/paypal/resteasy-spring-boot/blob/master/mds/USAGE.md#jax-rs-application-registration-methods), the starter won't do this classpath scanning by default. The default behavior is to have the JAX-RS Application subclass defined as a Spring bean, and rely on Spring to have it registered.

However, because dependency injection is not really part of the JAX-RS specification, we offer optionally registration of non Spring bean JAX-RS Application subclasses, and that is what this code is for. That is the third option by the way, a second one would be pointing to the class directly using a property.

By the way, when I implemented `JaxrsApplicationScanner`, I did look at what Spring could offer, and I did consider Spring `ClassPathScanningCandidateComponentProvider` as well, but I believe I noticed at that time that it only works with Spring beans. Isn't that true?

So, the bottom line is, classpath scanning feature is optional, and not recommended. Based on that, I believe I can close this as "not an issue". But let me know if I misunderstood something here, I would be glad to discuss it more. I will leave it open until you reply.

**Comment by snicoll** [comment]
I've looked at the code so I know classpath scanning is last resort. 

> I believe I noticed at that time that it only works with Spring beans. Isn't that true?

I gave you a working code that I actually used locally and all your tests were passing (it returns an object called a `BeanDefinition` but you can see in the code that we don't require the object to be a spring bean).

The only trick is that I scanned `""` (the root of the classpath) and you really don't want to do that. 

There are two reasons why I opened this issue:

1. Spring framework has a classpath scanning facility and we make sure that it works with all our scenario (including fat jar, etc). The code you have currently is looking for URLs and applies additional logic that could go away. If you have a dependency to spring (you do) it's worth a try
2. It would be better if you only scan what classspath scanning is configured for. I can only assume you scan everything right now and I don't think it is a good idea (even as a last ressort option).

**Comment by fabiocarvalho777** [comment]
> It would be more consistent that the actual value of component scan was used for the scanning (by default scan only the package in which the @SpringBootApplication is defined

That is a good idea. What is the best way to get a hold of that class? (without scanning the whole classpath, of course).

Also, there could be cases where this annotation is not used at all, for example:

``` java
@ComponentScan
@EnableAutoConfiguration
public class Application extends SpringBootServletInitializer {

	public static void main(String[] args) throws IOException {
		SpringApplication.run(Application.class);
	}

}

```
Is there a single and optimal way that would work for `SpringBootApplication` annotated and non-annotated classes?

**Comment by fabiocarvalho777** [comment]
> It would be more consistent that the actual value of component scan was used for the scanning (by default scan only the package in which the @SpringBootApplication is defined

That is a good idea. What is the best way to get a hold of that class? (without scanning the whole classpath, of course).

Also, there could be cases where this annotation is not used at all, for example:

``` java
@ComponentScan
@EnableAutoConfiguration
public class Application extends SpringBootServletInitializer {

	public static void main(String[] args) throws IOException {
		SpringApplication.run(Application.class);
	}

}

```
Is there a single and optimal way that would work for `SpringBootApplication` annotated and non-annotated classes?

**Comment by fabiocarvalho777** [comment]
[Here is unfinished solution for that](https://github.com/paypal/resteasy-spring-boot/commit/f802c36d74ba3e7d4375de4cecb6530821bbafe8). Just needs [this](https://github.com/paypal/resteasy-spring-boot/blob/issue_65/resteasy-spring-boot-starter/src/main/java/com/paypal/springboot/resteasy/ResteasyEmbeddedServletInitializer.java#L198-L201) to be completed.

**Comment by snicoll** [comment]
`AutoConfigurationPackages.get(BeanFactory beanFactory)` gives you that.

**Comment by fabiocarvalho777** [comment]
That gives me a list of packages. Is it safe to use just the first only? Or should I end up scanning all of them?

**Comment by snicoll** [comment]
No. Why would you ignore the other ones the user may have configured? I need to remember how it works (or you could try) but I think custom component scan directives are detected by this.

**Comment by fabiocarvalho777** [comment]
> Why would you ignore the other ones the user may have configured?

I don't mean to ignore anything. You pointed earlier in this issue that I should use the package where the `SpringBootApplication` annotated class is as the only place to be scanned, and then pointed me to `AutoConfigurationPackages.get(BeanFactory beanFactory)`, but it is clear that those two things are not the same.

In other words, it seems `AutoConfigurationPackages.get(BeanFactory beanFactory)` returns every single package to be scanned, when looking for beans to be registered. That scope is broader than implicit scanning for every bean defined under the main class package (which is what you suggested).

Anyways, I will just use that list of packages then, and get all of them scanned, looking for JAX-RS Application subclasses.

**Comment by fabiocarvalho777** [comment]
I am going to cover "scan every package registered by Spring framework, instead of the whole class-path" in this separated issue #69. The idea is to get that implemented only in the next major release (3.0.0), giving a chance for application to move away from it (in case they have their JAX-RS Application subclass scanned in a non Spring scanned package), and not breaking backward compatibility.

For this issue here, I am replacing the reflections based scanning by the Spring `ClassPathScanningCandidateComponentProvider` utility.

**Comment by fabiocarvalho777** [comment]
Done. See #70
